src/alloy.ts

```
@@ -1,45 +1,57 @@
 import { createInstance } from "@adobe/alloy";
 
-import { createAEPCtx } from "./contexts";
+import { AlloyInstance } from "./aep/types/alloy.types";
+import { isValidContext } from "./contexts/aep";
+import createContext from "./contexts/aep";
+import { BeaconSchema } from "./types/aep";
 import { AEPContext } from "./types/contexts";
 
-const alloy = createInstance({ name: "alloy" });
-
-function isValid(aep: AEPContext) {
-    return (
-        aep.datastreamId &&
-        aep.datastreamId !== "" &&
-        aep.imsOrgId &&
-        aep.imsOrgId !== ""
-    );
-}
-
-function configureAlloy(): Promise<any> {
-    const aepCtx: AEPContext = createAEPCtx();
-
-    return new Promise((resolve, reject) => {
-        if (isValid(aepCtx)) {
-            alloy("configure", {
-                edgeConfigId: aepCtx.datastreamId,
-                orgId: aepCtx.imsOrgId,
-                // TODO ahammond: remove debug when feature complete
+/*
+    We create a wrapper around `@adobe/alloy` for better type protection.
+*/
+
+const alloyInstance: AlloyInstance = createInstance({ name: "alloy" });
+
+/* configure alloy and assign it to the window object */
+const configure = async (): Promise<AlloyInstance | void> => {
```

These changes alter the "contract" of `configure`: originally the promise returned by `configure` would reject if the AEP context was invalid, but now it will resolve to `undefined`, which will result in an error further up the call stack.

I think we should either reject on invalid AEP context (preferred, since then we're failing fast) or make sure the rest of the code is robust enough to handle an `undefined` alloy instance.

---

i missed the reject there. let me refactor this a bit more, and i think it will solve your other comments as well